src/api/c/anisotropic_diffusion.cpp

```diff
@@ -0,0 +1,115 @@
+/*******************************************************
+ * Copyright (c) 2017, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/dim4.hpp>
+#include <af/image.h>
+#include <handle.hpp>
+#include <err_common.hpp>
+#include <backend.hpp>
+#include <arith.hpp>
+#include <cast.hpp>
+#include <copy.hpp>
+#include <gradient.hpp>
+#include <reduce.hpp>
+#include <anisotropic_diffusion.hpp>
+
+#include <type_traits>
+
+using af::dim4;
+using namespace detail;
+
+template<typename ComputeType, typename InputType, af_flux_function fftype, bool isMCDE>
+af_array diffusion(const af_array in, const float dt, const float K,
+                    const unsigned iterations)
+{
+    using T           = ComputeType;
+    using DiffusionFn = std::function< void(Array<T>&, const float, const float) >;
+
+    DiffusionFn diffuse = isMCDE ? curvatureDiffusion<T, fftype> : gradientDiffusion<T, fftype>;
+
+    auto out    = copyArray(castArray<T>(in));
+    auto dims   = out.dims();
+    auto g0     = createEmptyArray<T>(dims);
+    auto g1     = createEmptyArray<T>(dims);
+    float cnst  = -2.0*K*K/dims.elements();
+
+    for (unsigned i=0; i<iterations; ++i)
+    {
+        gradient<T>(g0, g1, out);
+
+        auto g0Sqr = arithOp<T, af_mul_t>(g0, g0, dims);
+        auto g1Sqr = arithOp<T, af_mul_t>(g1, g1, dims);
+        auto sumd  = arithOp<T, af_add_t>(g0Sqr, g1Sqr, dims);
+
+        T avg = reduce_all<af_add_t, T, T>(sumd, true, 0);
+
+        diffuse(out, dt, 1.0/(cnst*avg));
+    }
+
+    return getHandle(out);
+}
+
+template<af_flux_function fftype, bool isMCDE>
+af_err diffusion(af_array* out, const af_array in, const float dt,
+                 const float K, const unsigned iterations)
+{
+    try {
+        const ArrayInfo& info = getInfo(in);
+
+        const af::dim4& inputDimensions = info.dims();
+        const af_dtype  inputType       = info.getType();
+        const unsigned  inputNumDims    = inputDimensions.ndims();
+
+        DIM_ASSERT(1, (inputNumDims>=2));
+
+        ARG_ASSERT(4, (K>0 || K<0));
+
+        float DT = dt;
+        float maxDt = 1.0f/std::pow(2.0f, static_cast<float>(2)+1);
+        if (dt > maxDt) {
+            std::cerr<< "WARNING: maximum time step for 2D Image is "<< maxDt << std::endl;
```

Comment:
Not, this is not an error. ITK also doesn't prevent the user from running if the function if it was passed an unstable time step parameter. In our code, I added a debug build check to avoid printing anything in release mode.